File: pysubsonic.py
import uuid, requests, hashlib, json, pylast

class pySubsonic():
    def __init__(self, url, username, password, api, secret, lfm_user, lfm_password):
        self.url = url
        self.username = username
        self.password = password.decode()
        self.api = api.decode()
        self.secret = secret.decode()
        self.lfm_user = lfm_user
        self.lfm_password = lfm_password.decode()

    def _gen_salt(self, password):
        salt = uuid.uuid4().hex
        token = hashlib.md5((password + salt).encode()).hexdigest()
        return {"salt": salt, "token": token}

    def getPlaylists(self):
        gen = self._gen_salt(self.password)
        params = {
            'u': self.username,
            't': gen["token"],
            's': gen["salt"],
            'v': '1.16.1',
            'c': "MyApp",
            'f': "json"
        }

        response = requests.get(f"{self.url}/rest/getPlaylists.view", params=params)
        return response.json()
    
    def getSong(self, id):
        gen = self._gen_salt(self.password)
        params = {
            'u': self.username,
            't': gen["token"],
            's': gen["salt"],
            'v': '1.16.1',
            'c': "MyApp",
            'f': "json",
            'id': id
        }

        response = requests.get(f"{self.url}/rest/getSong.view", params=params)
        return response.json()["subsonic-response"]["song"]

    def getSongsInPlaylist(self, playlist):
        gen = self._gen_salt(self.password)
        params = {
            'u': self.username,
            't': gen["token"],
            's': gen["salt"],
            'v': '1.16.1',
            'c': "MyApp",
            'f': "json",
            'id': playlist
        }

        response = requests.get(f"{self.url}/rest/getPlaylist.view", params=params)
        return response.json()
    
    def songdetails(self, artist, title):
        pass_hash = pylast.md5(self.lfm_password)

        network = pylast.LastFMNetwork(
            api_key = self.api,
            api_secret = self.secret,
            username = self.username,
            password_hash = pass_hash
        )

        try:
            network.homepage
            return network.get_album(artist=artist, title=title)
        except pylast.WSError as e:
            print(f"Failed: {e}")
            return False
        
    def playlistCover(self, playlist):
        playlists = self.getPlaylists()["subsonic-response"]["playlists"]["playlist"]
        for item in playlists:
            if item["name"] == playlist:
                art_id = item["coverArt"]
        gen = self._gen_salt(self.password)
        return f"{self.url}/rest/getCoverArt.view?u={self.username}&t={gen["token"]}&s={gen["salt"]}&v=1.16.1&c=myClient&f=json&id={art_id}&size=300"
    
    def albumCover(self, song):
        gen = self._gen_salt(self.password)
        return f"{self.url}/rest/getCoverArt.view?u={self.username}&t={gen["token"]}&s={gen["salt"]}&v=1.16.1&c=myClient&f=json&id={song}&size=300"
        
    # songdetails(artist, title).get_cover_image(size=1)

File: subsonic-api/src/pySubsonic.js
const axios = require('axios');
const crypto = require('crypto');
const { v4: uuidv4 } = require('uuid');
const LastFm = require('lastfm');

class PySubsonic {
  constructor({ url, username, password, api, secret, lfmUser, lfmPassword }) {
    this.url = url;
    this.username = username;
    this.password = password;
    this.api = api;
    this.secret = secret;
    this.lfmUser = lfmUser;
    this.lfmPassword = lfmPassword;
  }

  // --- Private Helpers ---

  _genSalt() {
    const salt = uuidv4().replace(/-/g, '');
    const token = crypto
      .createHash('md5')
      .update(this.password + salt)
      .digest('hex');
    return { salt, token };
  }

  _baseParams() {
    const { salt, token } = this._genSalt();
    return {
      u: this.username,
      t: token,
      s: salt,
      v: '1.16.1',
      c: 'MyApp',
      f: 'json',
    };
  }

  async _get(endpoint, extraParams = {}) {
    const params = { ...this._baseParams(), ...extraParams };
    const response = await axios.get(`${this.url}/rest/${endpoint}`, { params });
    return response.data;
  }

  // --- Public Methods ---

  async getPlaylists() {
    return this._get('getPlaylists.view');
  }

  async getSong(id) {
    const data = await this._get('getSong.view', { id });
    return data['subsonic-response'].song;
  }

  async getSongsInPlaylist(playlist) {
    return this._get('getPlaylist.view', { id: playlist });
  }

  async songDetails(artist, title) {
    try {
      // Use 'lastfm' npm package instead of pylast
      const LastFm = require('lastfm-node-client');
      const lastfm = new LastFm(this.api, this.secret, this.lfmUser, this.lfmPassword);
      return await lastfm.albumGetInfo({ artist, album: title });
    } catch (err) {
      console.error(`Failed: ${err.message}`);
      return false;
    }
  }

  _coverArtUrl(id) {
    const { salt, token } = this._genSalt();
    const params = new URLSearchParams({
      u: this.username,
      t: token,
      s: salt,
      v: '1.16.1',
      c: 'myClient',
      f: 'json',
      id,
      size: '300',
    });
    return `${this.url}/rest/getCoverArt.view?${params.toString()}`;
  }

  async playlistCover(playlistName) {
    const data = await this.getPlaylists();
    const playlists = data['subsonic-response'].playlists.playlist;
    const match = playlists.find((p) => p.name === playlistName);
    if (!match) throw new Error(`Playlist "${playlistName}" not found`);
    return this._coverArtUrl(match.coverArt);
  }

  albumCover(songId) {
    return this._coverArtUrl(songId);
  }
}

module.exports = PySubsonic;